### PR TITLE
Refactor error page rendering and allow debug messages on error

### PR DESCRIPTION
**Pull request by JoelSpeed**
## Description

This moves the error page rendering to its own struct (with the required information captured) to break it apart from the main OAuth2 Proxy app.

This also adds a debug option so that detailed errors can be shown for users who choose this, on the error page.

## Motivation and Context

Improve the debug experience by rendering detailed errors in the web page when an authentication flow fails.

## How Has This Been Tested?

Manually with the local testing environment.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.
- [x] I have added appropriate comments throughout
- [x] I have added appropriate tests throughout

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Apart from the changelog entries, I've done all I want to here pre-review. Let me know what you think

File: pkg/apis/options/app.go
Comment (JoelSpeed) [inline pkg/apis/options/app.go:30]: I need to add this new flag to the docs... sigh, that list is too big!

File: pkg/app/error_page.go
Comment (NickMeves) [inline pkg/app/error_page.go:96]: Should this case use our same internal server error default message to keep things vague?

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:755]: I think this case is safe to not obfuscate and show the full error in the UI (debug or not) -- it is already part of the querystring in the request URL the enduser has if they got an error back from the IdP `authorization_endpoint`

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:762]: I think this message potentially gives too much details that a request is failing during the `token_endpoint` request with an invalid `authorization_code`

I think this & the CSRF check failing on a bad state should give the same generic error message to not tip an attacker off if they are attacking in some way as to which layer their attack failed.

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:790]: I think a CSRF cookie missing warrants the full error message right above this 👍 

But for this one, while I think knowing a CSRF failed is nice for debugging with an end user -- I'm torn on the opsec details an attacker gets if this and the code error above are different messages 🤔 

I recognize they are different today (even different response codes) But I think that is potentially an oversight we should remedy after seeing this more clearly.

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:766]: This `EnrichSessionState` call in general should be after the `state` CSRF protection.

Otherwise even if we align the error messages to be the same to not leak details, this is a noticeable network request in the middle -- so timing details would leak and tip our hand to attackers.

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:896]: I think this should mirror line 814 and be `http.StatusForbidden` (the underlying function returns this error in `!authorized` cases)

File: pkg/apis/options/app.go
Comment (NickMeves) [reply pkg/apis/options/app.go:30]: Hydra: Cut off one head two grow back in its place 😄 

It will only grow no matter how hard we try to slay it

**Review by NickMeves — CHANGES_REQUESTED** [review]
New error page handling logic & tests look great!

I have some concerns about the generic prod error messages we return that I'd like to discuss a bit before we merge this in though.

File: oauthproxy.go
Comment (JoelSpeed) [reply oauthproxy.go:762]: ToDo: Move to internal server error

File: oauthproxy.go
Comment (JoelSpeed) [reply oauthproxy.go:755]: ToDo: Create message var and duplicate the error string in both debug and non debug cases

File: oauthproxy.go
Comment (JoelSpeed) [reply oauthproxy.go:790]: ToDo: We have existing CSRF failures so leave this one as is for now until we resolve this later

File: oauthproxy.go
Comment (JoelSpeed) [reply oauthproxy.go:896]: ToDo: Make this consistent, should be `http.StatusForbidden`

File: pkg/app/error_page.go
Comment (JoelSpeed) [reply pkg/app/error_page.go:96]: This is a deliberately different error so that we can recognise missing `errorMessages`

**Review by NickMeves — APPROVED** [review]
LGTM!

As we discussed, once we figure out some of the underlying reasons for occasional CSRF cookie finicky-ness and get it more stable, we can make the changes around that logic.

**Comment by JoelSpeed** [comment]
I updated the docs and changelog after merge, and have checked those myself, will merge once tests pass

**Pull request merged**
